[4.2] Fix old input when submiting empty values

pxpm opened the pull request:
This is an old issue. Discussion about it here: #1816 

Fields not touched:

- checklist dependency
- enum (no sense)
- page_or_link
- select_from_array
- select_multiple
- select2_from_array
- select2_multiple
- view (no sense)

The ones I did not touch was mainly because they don't have fallbacks as the fields I added. Only followed field `default fallback` of the field that had them. Eg from text:

```php
value="{{ old(square_brackets_to_dots($field['name'])) ?? $field['value'] ?? $field['default'] ?? '' }}"
```
I fall back to:

```php
    if(is_null(old(square_brackets_to_dots($field['name']))) && !empty(session()->getOldInput())) {
        $field['value'] = '';
    }
```

scrutinizer-notifier commented:
A [new inspection](https://scrutinizer-ci.com/g/Laravel-Backpack/CRUD/inspections/dd532bf8-d086-4cdc-8df8-ebf25382dc3d) was created.

scrutinizer-notifier commented:
The [inspection](https://scrutinizer-ci.com/g/Laravel-Backpack/CRUD/inspections/dd532bf8-d086-4cdc-8df8-ebf25382dc3d) completed: **No new issues**

tabacitu commented:
```php

old('slug') = null => 0
backpack_old('slug', $fallback)

@php

    // now
    $field['value'] = old(square_brackets_to_dots($field['name'])) ?? $field['value'] ?? $field['default'] ?? 0 ;
    
    // proposal 1
    $field['value'] = oldOrFallback(square_brackets_to_dots($field['name']), 0) ?? $field['value'] ?? $field['default'] ?? 0 ;

    // proposal 2 - this would also work for date_range
    $field['value'] = oldOrFallback($field['name'], 0) ?? $field['value'] ?? $field['default'] ?? 0 ;
        // if validation error
            // if input was empty & 

    // proposal 3 - this would not work for date_range, for example
    $field['value'] = getFieldValue($field, 0);

    // proposal 4
    $field['value'] = oldOrFallback($field['name'], 0) ?? $field['value'] ?? $field['default'] ?? 0 ;

    
@endphp
```

Right now we agreed that Proposal 2 would probably be the best way to go forward. But the helper itself probably needs a better name.

tabacitu commented:
Note to self - this is still a thing. It's a known bug.

I've moved it to the 4.2 to-do-list. Since this changes so many fields, I think we'll be more comfortable with launching it as a breaking change.

pxpm commented:
Hello @tabacitu 

Sorry for the confusion here! I think you are right, but you were also rigth when we previously discussed this, so there is already #3054 

I've just merged master there, it addresses all the concerns you raised there too. 

Please have a look when you can and let me know.

I am going to close this one.

Pedro
